File: src/main/webapp/app/entities/reports/services/abc-pareto-report.service.ts
import { inject, Injectable } from '@angular/core';
import { HttpClient, HttpResponse, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';

import { SERVER_API_URL } from 'app/app.constants';
import { IABCPareto, IABCParetoSummary } from 'app/shared/model/report';
import { ClassePareto } from 'app/shared/model/report/classe-pareto.enum';

type EntityArrayResponseType = HttpResponse<IABCPareto[]>;
type EntityResponseType = HttpResponse<IABCParetoSummary>;

@Injectable({ providedIn: 'root' })
export class ABCParetoReportService {
  private readonly resourceUrl = SERVER_API_URL + 'api/abc-pareto';
  private readonly http = inject(HttpClient);

  /**
   * Get all ABC Pareto analysis data
   */
  getAllABCParetoAnalysis(): Observable<EntityArrayResponseType> {
    return this.http.get<IABCPareto[]>(this.resourceUrl, { observe: 'response' });
  }

  /**
   * Get ABC Pareto analysis filtered by category
   */
  getABCParetoByCategory(categorie: string): Observable<EntityArrayResponseType> {
    const params = new HttpParams().set('categorie', categorie);
    return this.http.get<IABCPareto[]>(`${this.resourceUrl}/category`, { params, observe: 'response' });
  }

  /**
   * Get ABC Pareto analysis filtered by Pareto class
   */
  getABCParetoByClass(classePareto: ClassePareto): Observable<EntityArrayResponseType> {
    const params = new HttpParams().set('classePareto', classePareto);
    return this.http.get<IABCPareto[]>(`${this.resourceUrl}/class`, { params, observe: 'response' });
  }

  /**
   * Get top N products by revenue contribution
   */
  getTopRevenueContributors(limit: number = 20): Observable<EntityArrayResponseType> {
    const params = new HttpParams().set('limit', limit.toString());
    return this.http.get<IABCPareto[]>(`${this.resourceUrl}/top`, { params, observe: 'response' });
  }

  /**
   * Get aggregated ABC Pareto summary
   */
  getABCParetoSummary(): Observable<EntityResponseType> {
    return this.http.get<IABCParetoSummary>(`${this.resourceUrl}/summary`, { observe: 'response' });
  }

  /**
   * Export ABC Pareto report as PDF
   */
  exportABCParetoToPdf(): Observable<HttpResponse<Blob>> {
    return this.http.get(`${this.resourceUrl}/export`, { observe: 'response', responseType: 'blob' });
  }
}

File: src/main/java/com/kobe/warehouse/service/report/ABCParetoReportService.java
package com.kobe.warehouse.service.report;

import com.kobe.warehouse.domain.enumeration.ClassePareto;
import com.kobe.warehouse.service.dto.report.ABCParetoDTO;
import com.kobe.warehouse.service.dto.report.ABCParetoSummaryDTO;
import java.util.List;

public interface ABCParetoReportService {
    /**
     * Get all ABC Pareto analysis data
     *
     * @return List of ABC Pareto records
     */
    List<ABCParetoDTO> getAllABCParetoAnalysis();

    /**
     * Get ABC Pareto analysis filtered by product category
     *
     * @param categorie The category name to filter by
     * @return List of ABC Pareto records for the category
     */
    List<ABCParetoDTO> getABCParetoByCategory(String categorie);

    /**
     * Get ABC Pareto analysis filtered by Pareto class
     *
     * @param classePareto The Pareto classification to filter by (A, B, C)
     * @return List of ABC Pareto records for the class
     */
    List<ABCParetoDTO> getABCParetoByClass(ClassePareto classePareto);

    /**
     * Get top N products by revenue contribution
     *
     * @param limit Number of products to return
     * @return List of top revenue contributors
     */
    List<ABCParetoDTO> getTopRevenueContributors(int limit);

    /**
     * Get aggregated ABC Pareto summary
     *
     * @return ABC Pareto summary with class distribution
     */
    ABCParetoSummaryDTO getABCParetoSummary();

    /**
     * Get ABC Pareto data with pagination
     *
     * @param page Page number (0-indexed)
     * @param size Page size
     * @return List of ABC Pareto records for the page
     */
    List<ABCParetoDTO> getABCParetoPaginated(int page, int size);

    /**
     * Get total count of ABC Pareto records
     *
     * @return Total count
     */
    long getABCParetoCount();

    /**
     * Get ABC Pareto by class with pagination
     *
     * @param classePareto The Pareto class to filter by
     * @param page Page number (0-indexed)
     * @param size Page size
     * @return List of ABC Pareto records for the page
     */
    List<ABCParetoDTO> getABCParetoByClassPaginated(ClassePareto classePareto, int page, int size);

    /**
     * Get count of ABC Pareto records by class
     *
     * @param classePareto The Pareto class to count
     * @return Count of records
     */
    long getABCParetoCountByClass(ClassePareto classePareto);
}
